File: functions/middleware/app-check.js
// functions/middleware/app-check.js
// Middleware para verificar Firebase App Check en Cloud Functions

const { createLogger } = require('../utils/structured-logger');
const logger = createLogger('app-check-middleware');

// Lazy require to avoid heavyweight import at module load
function getFunctions() {
  return require('firebase-functions/v1');
}

/**
 * Extraer información de versión del cliente desde headers
 * @param {Object} request - Express/Functions request object
 * @returns {Object} Información del cliente
 */
function extractClientInfo(request) {
  const headers = request?.headers || {};
  const userAgent = headers['user-agent'] || headers['User-Agent'] || '';
  
  // Intentar extraer versión desde headers personalizados
  let clientVersion = headers['x-client-version'] || headers['X-Client-Version'] || 
                      headers['x-app-version'] || headers['X-App-Version'] || null;
  
  // Si no hay header personalizado, intentar extraer de User-Agent
  if (!clientVersion && userAgent) {
    // Buscar versión de Firebase SDK en User-Agent
    // Formato típico: "firebase-js/10.12.2"
    const firebaseMatch = userAgent.match(/firebase-js\/([\d.]+)/i);
    if (firebaseMatch) {
      clientVersion = `firebase-js/${firebaseMatch[1]}`;
    }
  }
  
  return {
    clientVersion: clientVersion || 'unknown',
    userAgent: userAgent.substring(0, 200), // Limitar longitud
    ip: request?.ip || request?.connection?.remoteAddress || 'unknown'
  };
}

/**
 * Detectar si la solicitud parece venir de una versión antigua del SDK
 * @param {string} userAgent - User-Agent header
 * @param {string} clientVersion - Versión del cliente
 * @returns {boolean} True si parece ser legacy
 */
function detectLegacySDK(userAgent, clientVersion) {
  if (!userAgent && !clientVersion) {
    // Sin información de versión, asumir legacy
    return true;
  }
  
  if (clientVersion && clientVersion !== 'unknown') {
    // Si la versión del cliente es muy antigua, marcar como legacy
    // Versiones anteriores a 1.0.0 o sin formato de versión
    if (clientVersion.startsWith('0.') || !clientVersion.includes('/')) {
      return true;
    }
    
    // Detectar versiones antiguas de Firebase SDK (v8 y anteriores)
    if (clientVersion.startsWith('firebase-js/')) {
      const versionMatch = clientVersion.match(/firebase-js\/(\d+)/);
      if (versionMatch && parseInt(versionMatch[1]) < 9) {
        return true;
      }
    }
  }
  
  return false;
}

/**
 * Middleware para verificar App Check token en callable functions
 * @param {boolean} required - Si es true, rechaza requests sin App Check
 */
function requireAppCheck(required = true) {
  return (context) => {
    // App Check data está disponible en context.app
    // https://firebase.google.com/docs/app-check/cloud-functions

    const clientInfo = extractClientInfo(context.rawRequest);
    const isLegacy = detectLegacySDK(clientInfo.userAgent, clientInfo.clientVersion);
    
    if (!context.app) {
      logger.security('app_check_missing', {
        userId: context.auth?.uid,
        required,
        ip: clientInfo.ip,
        clientVersion: clientInfo.clientVersion,
        userAgent: clientInfo.userAgent,
        isLegacySDK: isLegacy,
        path: context.rawRequest?.path || 'callable',
        method: 'CALLABLE',
        timestamp: new Date().toISOString()
      });

      if (required) {
        throw new getFunctions().https.HttpsError(
          'failed-precondition',
          'App Check verification failed. Please update your app.'
        );
      }

      // Si no es requerido, solo logear advertencia con contexto completo
      logger.warn('Request without App Check allowed (not enforced)', {
        userId: context.auth?.uid,
        clientVersion: clientInfo.clientVersion,
        isLegacySDK: isLegacy,
        ip: clientInfo.ip
      });
    } else {
      logger.debug('App Check verified', {
        userId: context.auth?.uid,
        appId: context.app.appId,
        alreadyConsumed: context.app.alreadyConsumed,
        clientVersion: clientInfo.clientVersion,
        ip: clientInfo.ip
      });
    }
    
  };
}

/**
 * Middleware para verificar App Check en HTTP functions
 * @param {Object} req - Express request
 * @param {Object} res - Express response
 * @param {Function} next - Next middleware
 * @param {boolean} required - Si es true, rechaza requests sin App Check
 */
function verifyAppCheckHTTP(required = true) {
  return async (req, res, next) => {
    // Extraer información del cliente
    const clientInfo = extractClientInfo(req);
    const isLegacy = detectLegacySDK(clientInfo.userAgent, clientInfo.clientVersion);
    
    // Obtener información de autenticación si está disponible
    const authHeader = req.header('Authorization') || req.header('authorization') || '';
    const hasAuth = authHeader.startsWith('Bearer ');
    
    // Obtener App Check token del header
    const appCheckToken = req.header('X-Firebase-AppCheck');

    if (!appCheckToken) {
      // Logging estructurado con contexto completo
      logger.security('app_check_missing_http', {
        event: 'app_check_missing',
        type: isLegacy ? 'legacy_request' : 'missing_token',
        path: req.path,
        method: req.method,
        ip: clientInfo.ip,
        clientVersion: clientInfo.clientVersion,
        userAgent: clientInfo.userAgent,
        hasAuth: hasAuth,
        isLegacySDK: isLegacy,
        required,
        timestamp: new Date().toISOString()
      });

      if (required) {
        return res.status(401).json({
          error: 'unauthorized',
          message: 'App Check token missing'
        });
      }

      // Si no es requerido, continuar con advertencia con contexto completo
      logger.warn('HTTP request without App Check allowed (not enforced)', {
        path: req.path,
        clientVersion: clientInfo.clientVersion,
        isLegacySDK: isLegacy,
        ip: clientInfo.ip
      });
      return next();
    }

    try {
      // Verificar el token con Admin SDK
      const admin = require('firebase-admin');
      const appCheckClaims = await admin.appCheck().verifyToken(appCheckToken);

      // Logging estructurado para solicitudes exitosas
      logger.debug('App Check verified (HTTP)', {
        event: 'app_check_verified',
        appId: appCheckClaims.app_id,
        path: req.path,
        method: req.method,
        clientVersion: clientInfo.clientVersion,
        ip: clientInfo.ip,
        hasAuth: hasAuth,
        timestamp: new Date().toISOString()
      });

      // Agregar claims al request para uso posterior
      req.appCheckClaims = appCheckClaims;
      next();
    } catch (error) {
      // Logging estructurado para tokens inválidos
      logger.security('app_check_verification_failed', {
        event: 'app_check_invalid',
        path: req.path,
        method: req.method,
        error: error.message,
        errorType: error.constructor?.name || 'Unknown',
        ip: clientInfo.ip,
        clientVersion: clientInfo.clientVersion,
        userAgent: clientInfo.userAgent,
        hasAuth: hasAuth,
        isLegacySDK: isLegacy,
        timestamp: new Date().toISOString()
      });

      if (required) {
        return res.status(401).json({
          error: 'unauthorized',
          message: 'Invalid App Check token'
        });
      }

      // Si no es requerido, continuar con advertencia con contexto completo
      logger.warn('Invalid App Check token but continuing (not enforced)', {
        path: req.path,
        error: error.message,
        clientVersion: clientInfo.clientVersion,
        isLegacySDK: isLegacy,
        ip: clientInfo.ip
      });
      next();
    }
    
  };
}

module.exports = {
  requireAppCheck,
  verifyAppCheckHTTP
};

File: backend/app/middleware/app_check.py
"""
Middleware de verificación de Firebase App Check
Valida que las peticiones provengan de una instancia legítima de la aplicación.
Incluye logging estructurado y métricas para rastrear solicitudes legadas.
"""

from fastapi import Request, HTTPException, status
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import JSONResponse
import firebase_admin
from firebase_admin import auth, app_check
import logging
import os
from typing import Optional
from datetime import datetime

from app.utils.app_check_metrics import (
    get_metrics,
    detect_legacy_sdk,
    extract_client_version
)

logger = logging.getLogger("security.app_check")


def _extract_auth_token(request: Request) -> Optional[str]:
    """
    Extraer token de Firebase Auth del header Authorization si existe.
    
    Args:
        request: FastAPI Request object
        
    Returns:
        Token string o None
    """
    auth_header = request.headers.get("Authorization", "")
    if auth_header.startswith("Bearer "):
        return auth_header[7:]
    return None


def _get_user_id_from_token(token: Optional[str]) -> Optional[str]:
    """
    Intentar extraer user ID del token de Firebase Auth sin validar completamente.
    Solo para logging - no usar para autorización.
    
    Args:
        token: Firebase ID token
        
    Returns:
        User ID o None
    """
    if not token:
        return None
    
    try:
        # Decodificar sin verificar (solo para logging de métricas)
        import base64
        import json
        # JWT tiene formato: header.payload.signature
        parts = token.split('.')
        if len(parts) >= 2:
            # Decodificar payload
            payload = parts[1]
            # Agregar padding si es necesario
            padding = 4 - len(payload) % 4
            if padding != 4:
                payload += '=' * padding
            decoded = base64.urlsafe_b64decode(payload)
            claims = json.loads(decoded)
            return claims.get('uid') or claims.get('user_id')
    except Exception:
        # Si falla, no es crítico - solo para logging
        pass
    
    return None


class AppCheckMiddleware(BaseHTTPMiddleware):
    def __init__(self, app, exempt_paths: list = None):
        super().__init__(app)
        self.exempt_paths = exempt_paths or ["/docs", "/redoc", "/openapi.json", "/api/health", "/"]
        self.metrics = get_metrics()

    async def dispatch(self, request: Request, call_next):
        # 1. Verificar si la ruta está exenta
        path = request.url.path
        if any(path.startswith(exempt) for exempt in self.exempt_paths):
            return await call_next(request)
            
        # 1.5. Permitir OPTIONS (CORS Preflight)
        if request.method == "OPTIONS":
            return await call_next(request)
        
        # 2. Extraer información del cliente para logging y métricas
        client_host = request.client.host if request.client else "unknown"
        user_agent = request.headers.get("User-Agent", "")
        client_version = extract_client_version(dict(request.headers))
        auth_token = _extract_auth_token(request)
        user_id = _get_user_id_from_token(auth_token) if auth_token else None
        has_auth = bool(auth_token)
        
        # 3. Detectar si parece ser una versión antigua del SDK
        is_legacy = detect_legacy_sdk(user_agent, client_version)
        
        # 4. Verificar header X-Firebase-AppCheck
        app_check_token = request.headers.get("X-Firebase-AppCheck")
        
        if not app_check_token:
            # Logging estructurado para solicitudes sin App Check
            log_data = {
                "event": "app_check_missing",
                "type": "legacy_request" if is_legacy else "missing_token",
                "path": path,
                "method": request.method,
                "client_host": client_host,
                "client_version": client_version or "unknown",
                "user_agent": user_agent[:200] if user_agent else None,  # Limitar longitud
                "has_auth": has_auth,
                "user_id": user_id,
                "timestamp": datetime.utcnow().isoformat(),
                "is_legacy_sdk": is_legacy
            }
            
            logger.warning(
                f"Request without App Check token: {path} from {client_host}",
                extra={"log_data": log_data}
            )
            
            # Registrar en métricas
            self.metrics.record_request(
                has_app_check=False,
                client_version=client_version,
                path=path,
                is_legacy=is_legacy,
                token_valid=False
            )
            
            # SOFT FAIL - Permitir solicitud (Debug Mode)
            logger.info("⚠️ App Check Missing - Allowing request (Soft Enforcement)")
            return await call_next(request)
            
        try:
            # 5. Validar token con Firebase Admin SDK
            decoded_token = app_check.verify_token(app_check_token)
            app_id = decoded_token.get("app_id")
            
            # Logging estructurado para solicitudes exitosas
            log_data = {
                "event": "app_check_verified",
                "path": path,
                "method": request.method,
                "client_host": client_host,
                "client_version": client_version or "unknown",
                "app_id": app_id,
                "has_auth": has_auth,
                "user_id": user_id,
                "timestamp": datetime.utcnow().isoformat()
            }
            
            logger.debug(
                f"App Check verified: {path} (App ID: {app_id})",
                extra={"log_data": log_data}
            )
            
            # Registrar en métricas
            self.metrics.record_request(
                has_app_check=True,
                client_version=client_version,
                path=path,
                is_legacy=is_legacy,
                token_valid=True
            )
            
            # Continuar con la petición
            response = await call_next(request)
            return response
            
        except Exception as e:
            # Permitir fallo si es localhost o LAN para development
            is_dev = client_host in ["127.0.0.1", "::1", "localhost"] or client_host.startswith(("192.168.", "10."))
            
            if is_dev:
                logger.warning(
                    f"App Check failed in DEV/LAN ({client_host}), allowing access: {str(e)}",
                    extra={
                        "log_data": {
                            "event": "app_check_dev_bypass",
                            "path": path,
                            "client_host": client_host,
                            "error": str(e)
                        }
                    }
                )
                return await call_next(request)

            # Logging estructurado para tokens inválidos
            log_data = {
                "event": "app_check_invalid",
                "path": path,
                "method": request.method,
                "client_host": client_host,
                "client_version": client_version or "unknown",
                "user_agent": user_agent[:200] if user_agent else None,
                "has_auth": has_auth,
                "user_id": user_id,
                "error": str(e),
                "error_type": type(e).__name__,
                "timestamp": datetime.utcnow().isoformat(),
                "is_legacy_sdk": is_legacy
            }
            
            logger.error(
                f"Error validating App Check token: {str(e)}",
                extra={"log_data": log_data}
            )
            
            # Registrar en métricas
            self.metrics.record_request(
                has_app_check=True,  # Tenía token pero inválido
                client_version=client_version,
                path=path,
                is_legacy=is_legacy,
                token_valid=False
            )
            
            # SOFT FAIL - Permitir solicitud (Debug Mode)
            logger.info("⚠️ App Check Invalid - Allowing request (Soft Enforcement)")
            return await call_next(request)
